include/clap/ext/draft/audio-ports-activation.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include "../../plugin.h"
+
+/// @page Audio Ports Activation
+///
+/// This extension provides a way for the host to activate and de-activate audio ports.
+/// Deactivating a port provides the following benefits:
+/// - the plugin knows ahead of time that a given input is not present and can choose
+///   an optimized computation path
+/// - the plugin knows that an output is not consumed by the host, and doesn't need to
+///   compute it
+///
+/// Audio ports can only be activated or deactivated when the plugin is deactivated, unless
+/// can_activate_while_processing() returns true.
+///
+/// Audio buffers must still be provided if the audio port is deactivated.
+/// In such case, they shall be filled with 0 (or whatever is the neutral value in your context)
+/// and the constant_mask shall be set.
+///
+/// Audio ports are initially in the active state.
```

Comment:
It may also be worth explicitlymentioning whether or not the audio port state persists being deactivated and reactivated. I was just wondering about this and it seems to me like it would persist since it nowhere says that it does not, but it probably doesn't hurt to be explicit about it and I can imagine hosts or plugins interpreting this differently.

Reply:
I'm not sure if it should persist, but at least I'm sure that the host should explicitly set the state after loading the plugin, that'll avoid bugs.

Reply:
> I'm not sure if it should persist, but at least I'm sure that the host should explicitly set the state after loading the plugin, that'll avoid bugs.

What about compatibility here?
Say host-v1 doesn't support plugin port deactivation, it relies on plugins always having active ports.
Then host-v2 supports the extension, disables some ports and saves the state.
The project is shared with a system that runs host-v1. host-v1 loads the project/state, if the "deactivated" state persists, playback will be broken. So the state should definitely *not* persist.

So far I think the current draft is ill designed.
a) @abique may have a single use case in mind where port activation can be linked to a single user action and in a very limited scenario (plugin loading or deactivated engine) the port activation may be adjusted from the main thread.
b) However there are a number of cases (automatic routing, multi-channel support, noise gating, etc) where it'd really be useful to tell a plugin during process() on-the-fly if a specific port is currently in use or not. And that state is definitely temporary and for performance optimizations *only*.

The cases (b) can be all be solved with a new port activation event type, as can be case (a).
The current API proposal *only* supports case (a) though and is IMHO way more complex than a new event type.

Reply:
> ... So the state should definitely not persist.

Yes.

> b) ...

This is a good scenario but there are a bunch of implications here:
1. the plugin must be able to perform the activation / deactivation in a real-time context
1. DSP will have to communicate that state change back to the GUI
1. this optimization is difficult to implement both for the host and for the plugin, especially if it is an event which can happen at anytime (multiple time in a single process call); latency compensation also has to be considered. It could even be that in practice absolutely no one implements it.
1. it forces the plugin to pre-allocate all necessary buffers, while the current approach allows to only have what you need.
1. this design could also prevent some optimization, because it has a realtime semantic: "port (de)activation" happens at this given sample, while the current approach is more relaxed.

I don't buy this hardcore optimization arguments, and I doubt that any plugin vendor wants to deal with a port activation event which can happen in the middle of the process block.

While on the other hand, something that is more likely to be useful:
1. plugin gets the activation on the main thread
3. plugin needs to allocate some buffers and pre-compute some data
4. plugin passes this data in a lock free queue to its DSP

It is simpler because everything happens in the main thread all the time. In other plugin formats, ports activation only happens while the plugin is de-activated as far as I know, so it is already a huge improvement.

Reply:
I don't agree with 1) and 3) being problems at all. The whole point of this extension is to make changes in the plugin's audio processing. So it must already be able to handle these changes in a realtime context. And the plugin must also already be able to pass information from the process function to the GUI in a realtime safe way if it wants to have a GUI. The audio thread is the only place where the information is known to be up to date. Unless you add a mutex around the process funciton that can be locked from the GUI, in which case it would no longer be realtime safe.

Reply:
Here a proposal:
- this extension remain as it is
- we somehow find a way to tell the plugin if an output buffer is being produced/consumed, that could be done by setting the audio buffer to nullptr which'd be equivalent to "high impedance" state, if the plugin says that it supports this behavior.

Reply:
In the clap juce wrapper we are sort of assuming the buffers we ask for won't be null (I think surge checks anyway). So if we do this change of nullable buffers we would definitely want it opt in through the extension.

A "you are deactivated" main thread message which precedes the first "the buffer is now null" audio thread message makes sense to me though. But so does an event on or before the first block with a null buffer in the audio thread and audio has to tell gui.

but the one thing I surmise from this conversation is: if hosts are going to start using nullptr as a sentinel for deactivated buffers they should only do so with a host which has advertised the ability to handle that (through implementing this extension and advertising the appropriate handshake)

Hope that helps?

Reply:
Yes the handshake would be part of this extension, like we'd add `bool supports_null_audio_buffers()`.

Reply:
I'd have preferred to use a flag on the audio_buffer to indicate if it was computed or if it'll be consumed, but we don't have room for that.

Reply:
so we have 32 bits for channel count. I wonder if we could use a high bit there. but that is very bug prone. Since the constant mask is 64 bits we sort of have an implicit no-more-than-64-channel part of the API.

but null is best sentinel with the api we have I agree.

Reply:
Actually it is problematic to have two threads which can toggle the same value, or we'd need to have some kind of `AND`...
@robbert-vdh you won!